components/validator/challenger.go

```diff
@@ -24,232 +31,518 @@ type ProofFetcher interface {
 	if err != nil {
 		return nil, err
 	}
 
+	l2ooABI, err := bindings.L2OutputOracleMetaData.GetAbi()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(parentCtx)
 	callOpts := utils.NewSimpleCallOpts(ctx)
 	submissionInterval, err := l2ooContract.SUBMISSIONINTERVAL(callOpts)
 	if err != nil {
+		cancel()
 		return nil, fmt.Errorf("failed to get submission interval: %w", err)
 	}
 	finalizationPeriodSeconds, err := l2ooContract.FINALIZATIONPERIODSECONDS(callOpts)
 	if err != nil {
+		cancel()
 		return nil, fmt.Errorf("failed to get finalization period seconds: %w", err)
 	}
 	l2BlockTime, err := l2ooContract.L2BLOCKTIME(callOpts)
 	if err != nil {
+		cancel()
 		return nil, fmt.Errorf("failed to get l2 block time: %w", err)
 	}
 
 	return &Challenger{
-		log: l,
-		cfg: cfg,
-		ctx: ctx,
+		log:    l,
+		cfg:    cfg,
+		ctx:    ctx,
+		cancel: cancel,
 
 		callOpts: utils.NewCallOptsWithSender(ctx, cfg.TxManager.From()),
 		txOpts:   utils.NewSimpleTxOpts(ctx, cfg.TxManager.From(), cfg.TxManager.Signer),
 
+		l1Client: cfg.L1Client,
+
 		l2ooContract:      l2ooContract,
+		l2ooABI:           l2ooABI,
 		colosseumContract: colosseumContract,
+		colosseumABI:      colosseumABI,
 
 		submissionInterval:        submissionInterval,
 		finalizationPeriodSeconds: finalizationPeriodSeconds,
 		l2BlockTime:               l2BlockTime,
-	}, nil
-}
 
-func (c *Challenger) IsChallengeInProgress() (bool, error) {
-	return c.colosseumContract.IsInProgress(c.callOpts)
+		txCandidatesChan: txCandidateChan,
+	}, nil
 }
 
-func (c *Challenger) GetChallengeInProgress() (bindings.TypesChallenge, error) {
-	return c.colosseumContract.GetChallengeInProgress(c.callOpts)
-}
+func (c *Challenger) Start() error {
+	c.log.Info("start challenger")
 
-func (c *Challenger) OutputAtBlockSafe(blockNumber uint64, includeNextBlock bool) (*eth.OutputResponse, error) {
-	ctx, cancel := context.WithTimeout(c.ctx, c.cfg.NetworkTimeout)
-	defer cancel()
-	output, err := c.cfg.RollupClient.OutputAtBlock(ctx, blockNumber, includeNextBlock)
+	// if checkpoint is behind the latest output index, scan the previous outputs from the checkpoint
+	nextOutputIndex, err := c.l2ooContract.NextOutputIndex(c.callOpts)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("failed to get the latest output index: %w", err)
+	}
+	if nextOutputIndex.Cmp(common.Big0) == 0 {
+		// if no outputs have been submitted, set checkpoint to 1 because genesis output cannot be challenged.
+		c.checkpoint = common.Big1
+	} else {
+		// set checkpoint to latestOutputIndex (nextOutputIndex - 1)
+		c.checkpoint = new(big.Int).Sub(nextOutputIndex, common.Big1)
 	}
 
-	return output, nil
+	c.wg.Add(1)
+	go c.scanPrevOutputs()
+
+	// if challenge mode on, subscribe L2 output submission events
+	if !c.cfg.ChallengerDisabled {
+		c.wg.Add(1)
+		go c.subscribeL2OutputSubmitted()
+	}
+
+	// subscribe challenge creation events
+	c.wg.Add(1)
+	go c.subscribeChallengeCreated()
```

Comment:
`scanPrevOutputs()` and `subscribeChallengerCreated()` seem that those should be the function of the `Validator`, not `Challenger`.
Because those two should be called regardless of whether the challenger is enabled or not.

Reply:
I leave it as is for now, and will refactor for future.